pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/AbstractApexNode.java

```diff
@@ -57,4 +64,20 @@ public String getLocation() {
             return "no location";
         }
     }
+
+
+    @Override
+    public Iterator<Attribute> getXPathAttributesIterator() {
+        // Attributes of this node have precedence over same-name attributes of the underlying node
+        return Stream.concat(iteratorToStream(new AttributeAxisIterator(this)),
+                             iteratorToStream(new AttributeAxisIterator(this, node)))
```

Comment:
Hm... ok, so this exposes all the attributes/props of the underlying jorje node.

Naivly, I would have solved this problem by adding another method in ASTBooleanExpression that delegates the Operator and exposes it to be used.

Any changes to the attributes in the jorje lib will change the API of the node in PMD. I think, this will happen in any case (like the parse tree changes with an upgrade jorje lib or nodes appear/disappear). However, we would see e.g. missing/renamed attributes only at runtime (no compilation errors).

If we manually expose the attributes, we have the advantage, that the rules are coupled to PMD's API rather than to Jorje's API. E.g. now, if you want to write a Java rule that checks the operator, you'll need to know to look at the jorje node for the operator - it would be easier, if ASTBooleanExpression had a method "getOp" (now you need to call boolExpr.getNode().getOp()).
Of course, we will then need to expose any useful attribute manually...

I'm not sure whether we should suggest such a usage...

Reply:
You make a very interesting point, and I agree with it. We should aim to wrap Jorje and make sure we produce compilation errors if jorje changes (it usually does!) that way we can make sure to rework our internals with each jorje update and keep compatible outwards.

Only changes to the AST itself should impact users.